main/src/ca/uwaterloo/flix/language/phase/Namer.scala

```diff
@@ -1316,31 +1347,48 @@ object Namer extends Phase[WeededAst.Program, NamedAst.Root] {
   /**
     * Returns the implicit type parameters constructed from the given formal parameters and return type.
     */
-  private def getImplicitTypeParams(fparams: List[WeededAst.FormalParam], returnType: WeededAst.Type, loc: SourceLocation)(implicit flix: Flix): List[NamedAst.TypeParam] = {
+  private def getImplicitTypeParams(fparams: List[WeededAst.FormalParam], returnType: WeededAst.Type, loc: SourceLocation)(implicit flix: Flix): Validation[List[NamedAst.TypeParam], NameError] = {
     // Compute the type variables that occur in the formal parameters.
-    val typeVarsArgs = fparams.foldLeft(Set.empty[String]) {
-      case (acc, WeededAst.FormalParam(_, _, None, _)) => acc
-      case (acc, WeededAst.FormalParam(_, _, Some(tpe), _)) =>
-        freeVars(tpe).foldLeft(acc) {
-          case (innerAcc, ident) => innerAcc + ident.name
-        }
+    val typeVarsArgs = fparams.flatMap {
+      case WeededAst.FormalParam(_, _, Some(tpe), _) => freeVarsWithKind(tpe)
+      case WeededAst.FormalParam(_, _, None, _) => List.empty
     }
 
     // Compute the type variables that occur in the return type.
-    val typeVarsReturnType = freeVars(returnType).map(_.name)
+    val typeVarsReturnType = freeVarsWithKind(returnType)
 
     // Compute the set of type variables.
-    val typeVars = typeVarsArgs ++ typeVarsReturnType
+    val typeVars = typeVarsReturnType ::: typeVarsArgs
 
-    // Construct a (sorted) list of type parameters.
-    typeVars.toList.sorted.map {
-      case name =>
-        val ident = Name.Ident(SourcePosition.Unknown, name, SourcePosition.Unknown)
-        // We use a kind variable since we do not know the kind of the type variable.
-        val tvar = Type.freshVar(Kind.freshVar())
-        tvar.setText(name)
-        NamedAst.TypeParam(ident, tvar, loc)
+
```

Comment:
This is where the magic happens. I think we need some documentation to explain the purpose of the code and briefly the algorithm.

Reply:
I've added some documentation; let me know if there's somewhere that's not clear.